app/questionnaire/navigator.py

```diff
@@ -250,10 +260,76 @@ def get_latest_location(self, completed_blocks=None):
+
+    def get_front_end_navigation(self, completed_blocks, group_id, group_instance):
+        """
+        Returns the frontend navigation based on the completed blocks, group id and group instance
+
+        :param completed_blocks:
+        :param group_id:
+        :param group_instance:
+        :return: navigation
+        """
+        if 'navigation' not in self.survey_json:
+            return None
+
+        navigation = []
+
+        for group in filter(lambda x: 'hide_in_navigation' not in x, self.survey_json['groups']):
+
+            logger.debug("Building frontend navigation for group %s", group)
+
+            repeating_rule = SchemaHelper.get_repeating_rule(group)
+            completed_id = group['completed_id'] if 'completed_id' in group else group['blocks'][-1]['id']
+
+            if repeating_rule:
+                no_of_repeats = evaluate_repeat(repeating_rule, self.answer_store)
+                answer_id = repeating_rule['answer_id']
+                answers = self.answer_store.filter(answer_id=answer_id)
+
+                if repeating_rule['type'] == 'answer_count':
+                    self._add_repeating_navigation_item(answers, completed_blocks, completed_id, group, group_id,
+                                                        group_instance, navigation, no_of_repeats)
+                elif answers and no_of_repeats > 0:
+                    self._add_single_navigation_item(completed_blocks, completed_id, group, group_id, navigation)
+            else:
+                self._add_single_navigation_item(completed_blocks, completed_id, group, group_id, navigation)
+
+        return navigation
+
+    @staticmethod
+    def _add_repeating_navigation_item(answers, completed_blocks, completed_id, group, group_id, group_instance,
+                                       navigation, no_of_repeats):
+        for i in range(no_of_repeats):
+            if answers:
+                navigation.append({
+                    'link_name': answers[i]['value'],
+                    'group_id': group['id'],
+                    'instance': i,
+                    'block_id': group['blocks'][0]['id'],
+                    'completed': any(item for item in completed_blocks if item['group_instance'] == i and
+                                     item["block_id"] == completed_id),
+                    'highlight': group['id'] == group_id and i == group_instance,
+                    'repeating': True
+                })
+
+    @staticmethod
+    def _add_single_navigation_item(completed_blocks, completed_id, group, group_id, navigation):
+        navigation.append({
+            'link_name': group['title'],
+            'group_id': group['id'],
+            'instance': 0,
+            'block_id': group['blocks'][0]['id'],
+            'completed': any(item for item in completed_blocks if item["block_id"] == completed_id),
+            'highlight': (group['id'] != group_id and group_id in group['highlight_when']
```

Comment:
The highlight value specifically it may be clearer breaking it out into separate variables.